File: frontend/src/lib/services/conversations.ts
// Conversations Service
import { supabase } from '../supabase';
import type { Database } from '../supabase';

type Conversation = Database['public']['Tables']['conversations']['Row'];
type ConversationInsert = Database['public']['Tables']['conversations']['Insert'];
type ConversationUpdate = Database['public']['Tables']['conversations']['Update'];

export const conversationsService = {
    // Get all conversations for user
    async getConversations(userId: string) {
        const { data, error } = await supabase
            .from('conversations')
            .select('*')
            .eq('user_id', userId)
            .order('updated_at', { ascending: false });

        if (error) throw error;
        return data || [];
    },

    // Get conversation by ID
    async getConversationById(id: string) {
        const { data, error } = await supabase
            .from('conversations')
            .select('*')
            .eq('id', id)
            .single();

        if (error) throw error;
        return data;
    },

    // Create conversation
    async createConversation(conversation: Omit<ConversationInsert, 'id'>) {
        const { data, error } = await supabase
            .from('conversations')
            .insert(conversation)
            .select()
            .single();

        if (error) throw error;
        return data;
    },

    // Update conversation
    async updateConversation(id: string, updates: ConversationUpdate) {
        const { data, error } = await supabase
            .from('conversations')
            .update(updates)
            .eq('id', id)
            .select()
            .single();

        if (error) throw error;
        return data;
    },

    // Update conversation title
    async updateTitle(id: string, title: string) {
        return this.updateConversation(id, { title });
    },

    // Delete conversation (messages will be deleted via CASCADE)
    async deleteConversation(id: string) {
        const { error } = await supabase
            .from('conversations')
            .delete()
            .eq('id', id);

        if (error) throw error;
    },

    // Search conversations
    async searchConversations(userId: string, query: string) {
        const { data, error } = await supabase
            .from('conversations')
            .select('*')
            .eq('user_id', userId)
            .ilike('title', `%${query}%`)
            .order('updated_at', { ascending: false });

        if (error) throw error;
        return data || [];
    },
};

File: backend/app/routers/conversations.py
"""Conversations router — CRUD + messages."""

from fastapi import APIRouter, Depends, HTTPException

from app.supabase_client import get_supabase
from app.middleware.auth import get_current_user, AuthUser
from app.schemas.conversation import (
    ConversationCreate,
    ConversationResponse,
    MessageCreate,
    MessageResponse,
)

router = APIRouter()


@router.get("/", response_model=list[ConversationResponse])
async def list_conversations(user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    result = await sb.select("conversations", filters={"user_id": user.id}, order="-updated_at")
    return result or []


@router.post("/", response_model=ConversationResponse, status_code=201)
async def create_conversation(body: ConversationCreate, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    rows = await sb.insert("conversations", {"user_id": user.id, "title": body.title})
    if not rows:
        raise HTTPException(status_code=500, detail="Failed to create conversation")
    return rows[0]


@router.get("/{conversation_id}", response_model=ConversationResponse)
async def get_conversation(conversation_id: str, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    try:
        result = await sb.select("conversations", filters={"id": conversation_id, "user_id": user.id}, single=True)
    except Exception:
        raise HTTPException(status_code=404, detail="Conversation not found")
    if not result:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return result


@router.delete("/{conversation_id}", status_code=204)
async def delete_conversation(conversation_id: str, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    await sb.delete("conversations", {"id": conversation_id, "user_id": user.id})


# ── Messages ──────────────────────────────────────────────────────────────


@router.get("/{conversation_id}/messages", response_model=list[MessageResponse])
async def list_messages(conversation_id: str, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    # Verify ownership
    try:
        await sb.select("conversations", columns="id", filters={"id": conversation_id, "user_id": user.id}, single=True)
    except Exception:
        raise HTTPException(status_code=404, detail="Conversation not found")

    result = await sb.select("messages", filters={"conversation_id": conversation_id}, order="created_at")
    return result or []


@router.post("/{conversation_id}/messages", response_model=MessageResponse, status_code=201)
async def create_message(conversation_id: str, body: MessageCreate, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    try:
        await sb.select("conversations", columns="id", filters={"id": conversation_id, "user_id": user.id}, single=True)
    except Exception:
        raise HTTPException(status_code=404, detail="Conversation not found")

    rows = await sb.insert("messages", {
        "conversation_id": conversation_id,
        "role": body.role,
        "content": body.content,
        "provider": body.provider,
        "model": body.model,
    })
    if not rows:
        raise HTTPException(status_code=500, detail="Failed to save message")

    # Update conversation timestamp
    await sb.update("conversations", {"updated_at": "now()"}, {"id": conversation_id})

    return rows[0]

    return result.data[0]
